src/pubsub/streaming_subscription.rs

```diff
@@ -165,6 +204,209 @@ impl AcknowledgeToken {
     }
 }
 
+// Each ack token's ack/nack/modack calls are carried by this struct to the background task that
+// polls the mpsc channels. That task will then notify the token of the request's completion by sending
+// a result back over the given oneshot channel
+struct TokenFeedback<T> {
+    payload: T,
+    completion: oneshot::Sender<Result<(), AcknowledgeError>>,
+}
+
+impl<T: Send> TokenFeedback<T> {
+    fn send(
+        channel: &mpsc::UnboundedSender<TokenFeedback<T>>,
+        payload: T,
+    ) -> impl Future<Output = Result<(), AcknowledgeError>> + Send {
+        let (completion, listener) = oneshot::channel();
+
+        // send the payload over the channel synchronously. After this, the caller could drop the
+        // returned future and the work would still happen (barring errors/panics)
+        let send_result = channel.send(Self {
+            completion,
+            payload,
+        });
+
+        // now create the future to actually wait on the outcome
+        async move {
+            match send_result {
+                Ok(()) => match listener.await {
```

Comment:
When the oneshot::Receiver is used as a Future like here, does that mean it waits indefinitely? There's no need to timeout here?

Reply:
right, it waits until there is a response or until the sender half is dropped. A user can add a timeout around `ack()` in their application code if they'd like